# origin: Debian (herbert)
# cset: n/a
# inclusion: not submitted
# description: don't scan PCI bus until ide-probe-mod tells us to
# depends on: 082
# revision date: 2004-09-04

--- linux-2.4.27/drivers/ide/ide.c	2004-08-08 01:26:04.000000000 +0200
+++ linux-2.4.27+debian/drivers/ide/ide.c	2004-08-11 23:52:58.000000000 +0200
@@ -2489,12 +2489,9 @@
  */
 static void __init probe_for_hwifs (void)
 {
-#ifdef CONFIG_BLK_DEV_IDEPCI
-	if (pci_present())
-	{
-		ide_scan_pcibus();
-	}
-#endif /* CONFIG_BLK_DEV_IDEPCI */
+#if defined(CONFIG_BLK_DEV_IDEPCI) && !defined(MODULE)
+	ide_scan_pcibus();
+#endif /* CONFIG_BLK_DEV_IDEPCI && !MODULE */
 	ide_scan_drivers();
 
 	/*
--- linux-2.4.27/drivers/ide/setup-pci.c	2003-08-25 13:44:41.000000000 +0200
+++ linux-2.4.27+debian/drivers/ide/setup-pci.c	2004-08-11 23:52:58.000000000 +0200
@@ -793,9 +793,12 @@
 	struct pci_dev *dev;
 	struct pci_driver *d;
 	struct list_head *l, *n;
 
+	if (!pci_present() || !pre_init)
+		return;
+
 	pre_init = 0;
 	if (!ide_scan_direction) {
 		pci_for_each_dev(dev) {
 			ide_scan_pcidev(dev);
 		}
@@ -829,3 +832,5 @@
 		pci_register_driver(d);
 	}
 }
+
+EXPORT_SYMBOL_GPL(ide_scan_pcibus);
--- linux-2.4.27/include/linux/ide.h	2004-04-14 15:05:40.000000000 +0200
+++ linux-2.4.27+debian/include/linux/ide.h	2004-08-11 23:52:58.000000000 +0200
@@ -1749,6 +1749,8 @@
 typedef void (*ide_driver_call)(void);
 extern void __init ide_register_driver(ide_driver_call);
 
+extern void ide_driver_module(int revaldiate);
+
 /* ide locks for 2.4 */
 
 #define ide_lock		(io_request_lock)
